api/renter.go

```diff
@@ -350,49 +379,131 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
-			fileExists = true
+func (api *API) parseDownloadParameters(w http.ResponseWriter, req *http.Request, ps httprouter.Params) (*modules.RenterDownloadParameters, error) {
+	destination := req.FormValue("destination")
+
+	// The offset and length in bytes.
+	offsetparam := req.FormValue("offset")
+	lengthparam := req.FormValue("length")
+
+	// Determines whether the response is written to response body.
+	httprespparam := req.FormValue("httpresp")
+
+	// Determines whether to return on completion of download or straight away.
+	// If httprespparam is present, this parameter is ignored.
+	asyncparam := req.FormValue("async")
+
+	// Parse the offset and length parameters.
+	var offset, length uint64
+	var err error
+	if len(offsetparam) > 0 {
+		offset, err = strconv.ParseUint(offsetparam, 10, 64)
+		if err != nil {
+			return nil, build.ExtendErr("could not decode the offset as uint64: ", err)
 		}
 	}
+	if len(lengthparam) > 0 {
+		length, err = strconv.ParseUint(lengthparam, 10, 64)
+		if err != nil {
+			return nil, build.ExtendErr("could not decode the length as uint64: ", err)
+		}
+	}
+	// Verify that if either offset or length have been provided that both were provided.
+	offparampassed := len(offsetparam) > 0
+	lenparampassed := len(lengthparam) > 0
 
-	if !fileExists {
-		WriteError(w, Error{"file at specified siapath does not exist"}, http.StatusBadRequest)
-		return
+	// Parse the httpresp parameter.
+	httpresp, err := stringToBool(httprespparam)
+	if err != nil {
+		return nil, build.ExtendErr("httpresp parameter could not be parsed", err)
+	}
+
+	// Parse the async parameter.
+	async, err := stringToBool(asyncparam)
+	if err != nil {
+		return nil, build.ExtendErr("async parameter could not be parsed", err)
+	}
+
+	// Verify that either async or httpresp are true but not both.
+	if async && httpresp {
+		return nil, errors.New("only one of the async and httpresp flags can be specified.")
 	}
 
-	go api.renter.Download(strings.TrimPrefix(ps.ByName("siapath"), "/"), destination)
+	// Ensure that destination is defined if async is set to true.
+	if async && destination == "" {
+		return nil, errors.New("`destination` must be defined if `async=true`")
+	}
 
-	WriteSuccess(w)
+	// Ensure destination is not set if httpresp is defined.
+	if httpresp && destination != "" {
+		return nil, errors.New("`destination` cannot be defined if `httpresp=true`")
+	}
+
+	siapath := strings.TrimPrefix(ps.ByName("siapath"), "/") // Sia file name.
+
+	// Instantiate the correct DownloadWriter implementation
+	// (e.g. content written to file or response body).
+	var dw modules.DownloadWriter
+	if httpresp {
+		dw = modules.NewDownloadHttpWriter(w, offset, length)
+	} else {
+		// Ensure that destination is valid beforehands.
+		// Check that the destination path is absolute.
+		if !filepath.IsAbs(destination) {
+			return nil, &Error{"destination must be an absolute path"}
+		}
+		dw = modules.NewDownloadFileWriter(destination, offset, length)
```

Comment:
haven't gotten down to the code yet, but you probably actually want these things being created by the renter. The renter internals are going to have different optimizations for each type of writer, and you want the renter to be able to take clues and such to figure out the best way to structure the downloads.

e.g. httpwriter really doesn't want to use slow hosts, while filewriter wants to use as many hosts as it has memory to use.